src/core/xlio_types.h

```diff
@@ -0,0 +1,465 @@
+ * buffers. User can use this information to prepare the memory for some logic in the future.
+ * Zerocopy RX interface provides pointers to such memory.
+ *
+ * Argument hugepage_size provides the page size if XLIO uses hugepages for the allocation.
+ * If hugepage_size is not zero, the both addr and len are aligned to the page size boundary.
+ * There is no alignment guarantee for regular pages and hugepage_size is zero in this case.
+ * In case of external user allocator, XLIO reports hugepage_size zero regardless of the underlying
+ * pages properties.
+ */
+typedef void (*xlio_memory_cb_t)(void *addr, size_t len, size_t hugepage_size);
+
+/*
+ * Socket event callback.
+ *
+ * May be called from xlio_poll_group_poll() context.
+ * In the callback context, send operation is allowed only for the ESTABLISHED event.
+ * Argument value holds the error code for the ERROR event and 0 for other events.
+ *
+ * List of possible error code values:
+ * ECONNABORTED - connection aborted by local side
+ * ECONNRESET - connection reset by remote side
+ * ECONNREFUSED - connection refused by remote side during TCP handshake
+ * ETIMEDOUT - connection timed out due to keepalive, user timeout option or TCP handshake timeout
+ */
+enum {
+    /* TCP connection established. */
+    XLIO_SOCKET_EVENT_ESTABLISHED = 1,
+    /* Socket terminated and no further events are possible. */
+    XLIO_SOCKET_EVENT_TERMINATED,
+    /* Passive close. */
+    XLIO_SOCKET_EVENT_CLOSED,
+    /* An error occurred, see the error code value. */
+    XLIO_SOCKET_EVENT_ERROR,
+};
+typedef void (*xlio_socket_event_cb_t)(xlio_socket_t, uintptr_t userdata_sq, int event, int value);
+
+/*
+ * Zerocopy completion event.
+ *
+ * May be called from the following contexts:
+ *  - xlio_poll_group_poll() - likely
+ *  - xlio_socket_send() - can happen only if data is flushed
+ *  - xlio_socket_flush() / xlio_poll_group_flush()
+ *  - xlio_socket_destroy()
+ *
+ * In the callback context, send operation is allowed unless the socket is under destruction.
+ */
+typedef void (*xlio_socket_comp_cb_t)(xlio_socket_t, uintptr_t userdata_sq, uintptr_t userdata_op);
+
+/*
+ * RX callback.
+ *
+ * Returns TCP payload upon arrival. Each call returns a single contiguous buffer. The buffer points
+ * to memory within a block which is provided by the memory_cb() notification.
+ *
+ * xlio_buf is a descriptor of the buffer which must be returned to XLIO. During user ownership,
+ * they may use the uninitialized field in the structure.
+ */
+typedef void (*xlio_socket_rx_cb_t)(xlio_socket_t, uintptr_t userdata_sq, void *data, size_t len,
+                                    struct xlio_buf *buf);
+
+/*
+ * XLIO Socket API attribute structures
+ */
+
+struct xlio_init_attr {
+    unsigned flags;
+    xlio_memory_cb_t memory_cb;
+
+    /* Optional external user allocator for XLIO buffers. */
+    void *(*memory_alloc)(size_t);
```

Comment:
I believe this was a request to add external allocator. And it will  be used to allocate both TX and RX buffers. Is this OK that the same allocator used for both?

Reply:
This is done intentionally to simplify the default logic for user and XLIO. If differentiation between RX and TX memory becomes critical for some reason we will add a flag to `xlio_init_attr` to control that.